Move `TxIdentifier` into `bip152`

rustaceanrob opened the pull request:
Further crate de-coupling from `bitcoin`.

~This generalizes the API for something `TxIdentifier` implements without having to duplicate type. If/when `TxIdentifier` is public in primitives, we can replace it again.~

Move the trait from `bitcoin/blockdata`, where it was only being used via `p2p`.

tcharding commented:
Since, in this crate, the trait is exactly an abstraction that governs this single function I think it should be duplicated. But I'm not going to die on this hill :)

mpbagot commented:
I don't think `TxIdentifier` does anything in `bitcoin` anymore. As far as I can see, it's only defined + exported from `bitcoin` and then imported + used in `p2p`. You could just move it from `bitcoin` to `p2p` entirely.

While I personally think it'd be better to make it pub from `primitives`, @apoelstra specifically shot that down in https://github.com/rust-bitcoin/rust-bitcoin/pull/5346#issuecomment-3591702046. I'm not sure what the reasoning was though, beyond just a desire to minimise public API changes in `primitives`.

rustaceanrob committed:
Move `TxIdentifier` into `bip152`

This trait was used in `bitcoin/blockdata/transaction` only to be used
in `p2p`. Move it to `p2p` and drop it in `bitcoin`. I opted to include
it directly in BIP-152, unless another context arises where these IDs
are interchangable.

Credit @mpbagot for the find

tcharding commented:
> While I personally think it'd be better to make it pub from primitives, @apoelstra specifically shot that down

FWIW the reasoning is that we are trying to keep `primitives 1.0.0` and the other two 1.0 crates to have as small an API as possible. We spent close to two years teasing `units` and `primitives` out of `bitcoin` trying to achieve this goal - it was brutal.

apoelstra commented:
Needs rebase after #5425.

The reasoning is not just about API size. It's that this trait is a terrible abstraction.

tcharding reviewed (dismissed):
ACK 250a027a716800a4cf09944bd685a4b0b9601499

tcharding commented on p2p/src/bip152.rs:
```diff
@@ -93,6 +92,17 @@ impl Decodable for PrefilledTransaction {
     }
 }
+/// Trait that abstracts over a transaction identifier i.e., `Txid` and `Wtxid`.
+pub trait TxIdentifier: sealed::Sealed + AsRef<[u8]> {}
```
nit: Missing new line above doc line.

apoelstra reviewed (approved):
ACK b7f035be01b58a744449750723e05867a07b5a63; successfully ran local tests

tcharding reviewed (approved):
ACK b7f035be01b58a744449750723e05867a07b5a63
